### PR TITLE
Feature/edit washer

### PR DESCRIPTION
**Pull request by joelgroomer**
# Description

Allows Washers to edit their pricing details, "about me", and set their work status.

Still need to activate the map and also report the current location back to the server.

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Change Status

- [x] Complete, tested, ready to review and merge

# How Has This Been Tested?

- [x] Manual testing
- [x] Automated tests

# Checklist

- [x] My code follows the style guidelines of this project
- [ x I have performed a self-review of my own code
- [ ] My code has been reviewed by at least one peer
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] There are no merge conflicts

### CODE REVIEW

File: ExpressWash/ExpressWash/Base.lproj/Main.storyboard [hunk 1]
```diff
@@ -898,7 +898,7 @@
                         <outlet property="jobDescriptionLabel" destination="6GO-nA-ecJ" id="xht-Xg-vPt"/>
                         <outlet property="licPlateLabel" destination="hdf-SL-Yu6" id="uMM-51-aGC"/>
                         <outlet property="makeModelLabel" destination="NNY-mJ-z24" id="hfa-Hf-p7y"/>
-                        <outlet property="mapView" destination="dYN-CP-n1b" id="oGh-Z9-OWl"/>
+                        <outlet property="mapView" destination="dYN-CP-n1b" id="ADh-0Z-WLk"/>
                         <outlet property="profileImageView" destination="NbW-MF-ct6" id="m41-Ix-FHN"/>
                         <outlet property="rateLargeLabel" destination="Cbi-Km-kyC" id="JWs-VA-Ip9"/>
                         <outlet property="rateMediumLabel" destination="F8h-RP-QrO" id="dFR-6I-DP0"/>
```
Comment (Keffury1) [inline ExpressWash/ExpressWash/Base.lproj/Main.storyboard:904]: Pretty cool how easy the map was to implement.

File: ExpressWash/ExpressWash/EditWasherViewController.swift [hunk 2]
```diff
@@ -17,28 +17,33 @@ class EditWasherViewController: UIViewController {
     @IBOutlet weak var txtRateLarge: UITextField!
     @IBOutlet weak var txtRateMedium: UITextField!
     @IBOutlet weak var txtRateSmall: UITextField!
+    @IBOutlet weak var btnSave: UIButton!
 
     // MARK: - Properties
-    var washer: Washer? { didSet { updateViews() } }
+//    var washer: Washer? { didSet { updateViews() } }
```
Comment (Keffury1) [inline ExpressWash/ExpressWash/EditWasherViewController.swift:23]: is this needed? if so, uncomment eventually

File: ExpressWash/ExpressWash/EditWasherViewController.swift [hunk 3]
```diff
@@ -17,28 +17,33 @@ class EditWasherViewController: UIViewController {
...
     }
 
     func updateViews() {
         guard isViewLoaded else { return }
 
-        guard let washer = washer,
-            let user = washer.user else { return }
+        guard let washer = UserController.shared.sessionUser.washer,
+            let user = washer.user else {
+                btnSave.isEnabled = false
```
Comment (Keffury1) [inline ExpressWash/ExpressWash/EditWasherViewController.swift:38]: Good logic.

File: ExpressWash/ExpressWash/EditWasherViewController.swift [hunk 4]
```diff
@@ -17,28 +17,33 @@ class EditWasherViewController: UIViewController {
...
+                return
+        }
 
+        btnSave.isEnabled = true
         lblFullName.text = "\(user.firstName) \(user.lastName)"
```
Comment (Keffury1) [inline ExpressWash/ExpressWash/EditWasherViewController.swift:43]: Capitalized perhaps?

File: ExpressWash/ExpressWash/EditWasherViewController.swift [hunk 5]
```diff
@@ -53,9 +58,53 @@ class EditWasherViewController: UIViewController {
...
+            let smallStr = txtRateSmall.text,
+            let small = Double(smallStr)
+        else {
+            saveAlert(msg: "Please enter a large, medium, and small car rate. Use only numbers and decimal points.")
+            return
+        }
+        washerRep.rateLarge = large
+        washerRep.rateMedium = medium
+        washerRep.rateSmall = small
+        // TODO: Get current location from phone and assign it to the rep
```
Comment (Keffury1) [inline ExpressWash/ExpressWash/EditWasherViewController.swift:86]: Todo here.

File: ExpressWash/ExpressWash/EditWasherViewController.swift [hunk 6]
```diff
@@ -53,9 +58,53 @@ class EditWasherViewController: UIViewController {
...
+        self.dismiss(animated: true, completion: nil)
     }
 
+    private func saveAlert(msg: String) {
+        let alert = UIAlertController()
+        alert.title = "Error"
+        alert.message = msg
+        alert.addAction(UIAlertAction(title: "OK", style: .default, handler: nil))
+        self.present(alert, animated: true, completion: nil)
+    }
```
Comment (Keffury1) [inline ExpressWash/ExpressWash/EditWasherViewController.swift:109]: Pretty standard. Looks good

File: ExpressWash/ExpressWash/SessionUser.swift [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+            self.washer = nil
+        }
+    }
+
+    func encode(to encoder: Encoder) throws {
+        var container = encoder.container(keyedBy: SessionUserKeys.self)
+        try container.encode(self.user?.representation, forKey: .user)
+        try container.encode(self.washer?.representation, forKey: .washer)
+    }
+}
```
Comment (Keffury1) [inline ExpressWash/ExpressWash/SessionUser.swift:43]: Sweet that you got this figured out at running. looks clean.

File: ExpressWash/ExpressWash/UserController+authenticate.swift [hunk 8]
```diff
@@ -62,13 +62,28 @@ extension UserController {
...
+
+                    // find the Washer in Core Data, if it exists
+                    self.sessionUser.washer = washerController.findWasher(byID: washerRep.washerId)
+                    if self.sessionUser.washer != nil {
+                        // if it did exist, update it with the newly received data
+                        washerController.updateWasher(self.sessionUser.washer!, with: washerRep)
+                    } else {
+                        // if it didn't exist, create it and save it to the session
+                        self.sessionUser.washer = Washer(representation: washerRep)
+                    }
```
Comment (Keffury1) [inline ExpressWash/ExpressWash/UserController+authenticate.swift:84]: Comments are solid here

File: ExpressWash/ExpressWash/WasherViewController.swift [hunk 9]
```diff
@@ -141,18 +149,25 @@ class WasherViewController: UIViewController, MGLMapViewDelegate {
...
 
         var washerRep = washer.representation
         washerRep.workStatus = activeSwitch.isOn
-        // TODO: Get current location from phone and assign it to the rep
+        if washerRep.workStatus {
+            locationManager?.requestLocation()
+            if let lat = locationManager?.location?.coordinate.latitude,
+                let lon = locationManager?.location?.coordinate.longitude {
+                washerRep.currentLocationLat = lat
+                washerRep.currentLocationLon = lon
```
Comment (Keffury1) [inline ExpressWash/ExpressWash/WasherViewController.swift:168]: Great work here

File: ExpressWash/ExpressWashTests/Info.plist [hunk 10]
```diff
@@ -18,6 +20,8 @@
 	<string>1.0</string>
 	<key>CFBundleVersion</key>
 	<string>1</string>
+	<key>NSLocationWhenInUseUsageDescription</key>
+	<string>We use your location to match up people who need a wash with washers who are nearby.</string>
```
Comment (Keffury1) [inline ExpressWash/ExpressWashTests/Info.plist:24]: Great catch on the testing plist

**Review by Keffury1 — APPROVED** [review]
Looks like solid work. Mostly fixing the sessionUser issue and correcting it accordingly.

File: ExpressWash/ExpressWash/EditWasherViewController.swift (hunk 2 above)
Comment (joelgroomer) [reply ExpressWash/ExpressWash/EditWasherViewController.swift:23]: It's not. Just missed deleting it. Thanks!

**Pull request merged**